src/main/groovy/org/boozallen/plugins/jte/init/primitives/TemplatePrimitive.groovy

```diff
@@ -16,57 +16,63 @@
 package org.boozallen.plugins.jte.init.primitives
 
 import com.cloudbees.groovy.cps.NonCPS
+import org.boozallen.plugins.jte.util.JTEException
+import org.boozallen.plugins.jte.util.TemplateLogger
+import org.jenkinsci.plugins.workflow.cps.CpsScript
+import org.jenkinsci.plugins.workflow.cps.GlobalVariable
+
+import javax.annotation.Nonnull
 
 /**
- * Objects whose class extends TemplatePrimitive will be protected in the {@link TemplateBinding} from
- * being inadvertently overridden
+ * Framework constructs that make templates easier to write.
+ * Typically created by parsing the Pipeline Configuration.
  */
-abstract class TemplatePrimitive implements Serializable{
+abstract class TemplatePrimitive extends GlobalVariable implements Serializable{
 
     private static final long serialVersionUID = 1L
 
-    Class<? extends TemplatePrimitiveInjector> injector
-    String name
-
-    @NonCPS
-    Object getValue(){ return this }
-
     /**
-     * Invoked if an object with this class were to be overridden in the {@link TemplateBinding} during initialization
+     * The GlobalCollisionValidator will populate this list with all
+     * TemplatePrimitives sharing the same name if there is
+     * more than 1.
      */
-    abstract void throwPreLockException(String preface)
+    List<TemplatePrimitive> overloaded = []
 
-    /**
-     * Invoked if an object with this class were to be overridden in the {@link TemplateBinding} after initialization
-     */
-    abstract void throwPostLockException(String preface)
+    String name
+    TemplatePrimitiveNamespace parent
 
-    /**
-     * Returns the injector that creates the primitive
-     * <p>
-     * implementing classes must mark this method @NonCPS lest a CpsCallableInvocation be thrown
-     * during initialization
-     * @return
-     */
-    abstract Class<? extends TemplatePrimitiveInjector> getInjector()
+    @Override
+    @NonCPS
+    Object getValue(@Nonnull CpsScript script) throws Exception {
```

Comment:
maybe refactor the overload check and exception into another method that is reused by subclasses